File: intelligence/harvester.py
"""
Amazon KDP Intelligence Harvester — REFERENCE ONLY
====================================================
WARNING: This script uses requests + BeautifulSoup for HTTP scraping.
Amazon detects and blocks this approach (CAPTCHA / 503 responses) almost immediately.

THE ACTUAL HARVEST METHOD is the harvester-agent.md agent, which uses
Playwright's browser_evaluate() in a real browser session. This Python file
is kept as a reference for the data structures and DB-write logic only.

DO NOT call this script expecting it to successfully harvest Amazon data.
Use: `harvest [niche]` to invoke the real harvester-agent via Playwright.

If you need to load pre-scraped data from a file, see load_to_db.py instead.

Original run intent: python intelligence/harvester.py [niche] [marketplace]
"""

import sqlite3
import json
import time
import re
import sys
from datetime import datetime
from pathlib import Path

try:
    import requests
    from bs4 import BeautifulSoup
except ImportError:
    print("Installing required packages...")
    import subprocess
    subprocess.check_call([sys.executable, "-m", "pip", "install", "requests", "beautifulsoup4", "lxml"])
    import requests
    from bs4 import BeautifulSoup

DB_PATH = Path(__file__).parent / "opportunity.db"
NICHES_PATH = Path(__file__).parent / "niches.json"

HEADERS = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36",
    "Accept-Language": "en-GB,en;q=0.9",
    "Accept": "text/html,application/xhtml+xml,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
    "Accept-Encoding": "gzip, deflate, br",
    "DNT": "1",
    "Connection": "keep-alive",
    "Upgrade-Insecure-Requests": "1",
}

MARKETPLACES = {
    "UK": "https://www.amazon.co.uk",
    "US": "https://www.amazon.com",
}


def get_page(url: str, session: requests.Session, retries: int = 2) -> BeautifulSoup | None:
    for attempt in range(retries):
        try:
            resp = session.get(url, headers=HEADERS, timeout=15)
            if resp.status_code == 200:
                return BeautifulSoup(resp.text, "lxml")
            if resp.status_code == 503:
                print(f"  [!] Amazon bot check on attempt {attempt+1}. Waiting...")
                time.sleep(5 * (attempt + 1))
        except Exception as e:
            print(f"  [!] Request error: {e}")
            time.sleep(3)
    return None


def parse_price(text: str) -> float | None:
    if not text:
        return None
    match = re.search(r"[\d,]+\.?\d*", text.replace(",", ""))
    return float(match.group()) if match else None


def parse_int(text: str) -> int | None:
    if not text:
        return None
    match = re.search(r"[\d,]+", text.replace(",", ""))
    return int(match.group().replace(",", "")) if match else None


def scrape_search_page(soup: BeautifulSoup) -> list:
    products = []
    for el in soup.select("[data-asin]"):
        asin = el.get("data-asin", "")
        if not asin or len(asin) < 5:
            continue
        title_el = el.select_one("h2 span")
        price_el = el.select_one(".a-price .a-offscreen")
        rating_el = el.select_one("[aria-label*='out of 5 stars']")
        review_el = el.select_one(".a-size-base.s-underline-text")
        ku = bool(el.select_one(".a-badge-supplemental, [aria-label*='Kindle Unlimited']"))

        title = title_el.get_text(strip=True) if title_el else None
        if not title:
            continue

        raw_reviews = review_el.get_text(strip=True) if review_el else ""
        review_count = parse_int(raw_reviews) if re.search(r"[\d,]+", raw_reviews) else None
        rating_text = rating_el.get("aria-label", "") if rating_el else ""
        rating_match = re.search(r"([\d.]+) out of", rating_text)
        rating = float(rating_match.group(1)) if rating_match else None

        products.append({
            "asin": asin,
            "title": title,
            "price_ebook": parse_price(price_el.get_text() if price_el else ""),
            "star_rating": rating,
            "review_count": review_count,
            "on_ku": ku,
        })
    # Deduplicate by ASIN
    seen = set()
    unique = []
    for p in products:
        if p["asin"] not in seen:
            seen.add(p["asin"])
            unique.append(p)
    return unique


def scrape_product_bsr(asin: str, base_url: str, session: requests.Session) -> dict:
    url = f"{base_url}/dp/{asin}"
    soup = get_page(url, session)
    if not soup:
        return {}

    result = {}

    # BSR from detail bullets
    bsr_data = []
    for li in soup.select("#detailBullets_feature_div li"):
        text = li.get_text(" ", strip=True)
        if "#" in text and ("Kindle" in text or "Books" in text or "Store" in text):
            bsr_data.append(text)

    if bsr_data:
        # Parse first BSR entry (overall)
        first = bsr_data[0]
        overall_match = re.search(r"#([\d,]+) in", first)
        if overall_match:
            result["bsr_overall"] = parse_int(overall_match.group(1))
        # Category name
        cat_match = re.search(r"in (.+?)(?:\s*\(|$)", first)
        if cat_match:
            result["bsr_category1"] = cat_match.group(1).strip()
        if len(bsr_data) > 1:
            second = bsr_data[1]
            rank2 = re.search(r"#([\d,]+)", second)
            cat2 = re.search(r"in (.+?)(?:\s*\(|$)", second)
            if rank2:
                result["bsr_category2_rank"] = parse_int(rank2.group(1))
            if cat2:
                result["bsr_category2"] = cat2.group(1).strip()

    # Pages
    for li in soup.select("#detailBullets_feature_div li"):
        text = li.get_text(" ", strip=True)
        if "print length" in text.lower() or "pages" in text.lower():
            pages_match = re.search(r"([\d,]+)\s*pages", text, re.IGNORECASE)
            if pages_match:
                result["page_count"] = parse_int(pages_match.group(1))
            break

    # Publication date
    for li in soup.select("#detailBullets_feature_div li"):
        text = li.get_text(" ", strip=True)
        if "publication date" in text.lower() or "date first" in text.lower():
            result["pub_date"] = text.split(":")[-1].strip()
            break

    # KU check
    result["on_ku"] = bool(soup.select_one("[data-action='kindle-unlimited-button']"))

    # Kindle price
    price_el = soup.select_one(".kindle-price .a-price .a-offscreen, #kindle-price")
    if price_el:
        result["price_ebook"] = parse_price(price_el.get_text())

    # Review count
    review_el = soup.select_one("#acrCustomerReviewText")
    if review_el:
        result["review_count"] = parse_int(review_el.get_text())

    # Rating
    rating_el = soup.select_one("#acrPopover")
    if rating_el:
        rating_match = re.search(r"([\d.]+) out of", rating_el.get("title", ""))
        if rating_match:
            result["star_rating"] = float(rating_match.group(1))

    return result


def save_products(products: list, niche: str, marketplace: str):
    conn = sqlite3.connect(DB_PATH)
    saved = 0
    now = datetime.now().isoformat()
    for p in products:
        try:
            conn.execute("""
                INSERT OR IGNORE INTO products
                (asin, marketplace, niche, title, author, price_ebook, price_paperback,
                 bsr_overall, bsr_category1, bsr_category1_rank, bsr_category2,
                 bsr_category2_rank, review_count, star_rating, pub_date,
                 page_count, on_ku, scraped_at, raw_json)
                VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)
            """, (
                p.get("asin"), marketplace, niche,
                p.get("title"), p.get("author"),
                p.get("price_ebook"), p.get("price_paperback"),
                p.get("bsr_overall"),
                p.get("bsr_category1"), p.get("bsr_category1_rank"),
                p.get("bsr_category2"), p.get("bsr_category2_rank"),
                p.get("review_count"), p.get("star_rating"),
                p.get("pub_date"), p.get("page_count"),
                1 if p.get("on_ku") else 0,
                now, json.dumps(p)
            ))
            saved += 1
        except Exception as e:
            print(f"  [DB] Skip {p.get('asin')}: {e}")
    conn.commit()
    conn.close()
    return saved


def harvest_niche(niche_config: dict, marketplace: str):
    name = niche_config["name"]
    base_url = MARKETPLACES[marketplace]
    search_terms = niche_config["search_terms"]

    print(f"\n{'='*50}")
    print(f"Harvesting: {name} | {marketplace}")
    print(f"{'='*50}")

    session = requests.Session()
    all_products = {}

    for term in search_terms[:3]:  # Top 3 search terms per niche
        encoded = term.replace(" ", "+")
        url = f"{base_url}/s?k={encoded}&i=digital-text"
        print(f"\n  Searching: '{term}'")
        soup = get_page(url, session)
        if not soup:
            print(f"  [!] Could not fetch results for '{term}'")
            continue

        products = scrape_search_page(soup)
        print(f"  Found {len(products)} products")
        for p in products:
            if p["asin"] not in all_products:
                all_products[p["asin"]] = p

        time.sleep(2)  # Be polite

    print(f"\n  Total unique products: {len(all_products)}")

    # Get BSR for top 20 products
    product_list = list(all_products.values())
    print(f"\n  Getting BSR for top {min(20, len(product_list))} products...")

    for i, p in enumerate(product_list[:20]):
        print(f"  [{i+1}/20] {p['asin']} - {p['title'][:50]}...")
        bsr_data = scrape_product_bsr(p["asin"], base_url, session)
        p.update(bsr_data)
        print(f"    BSR: {p.get('bsr_overall', 'N/A')} | Reviews: {p.get('review_count', 'N/A')} | KU: {p.get('on_ku', False)}")
        time.sleep(1.5)

    saved = save_products(product_list, name, marketplace)
    print(f"\n  Saved {saved} products to database")
    return saved


def main():
    niches = json.loads(NICHES_PATH.read_text())["niches"]

    target_niche = sys.argv[1] if len(sys.argv) > 1 else None
    target_marketplace = sys.argv[2] if len(sys.argv) > 2 else None

    total_saved = 0
    for niche in niches:
        if target_niche and niche["name"] != target_niche:
            continue
        for marketplace in niche["marketplaces"]:
            if target_marketplace and marketplace != target_marketplace:
                continue
            saved = harvest_niche(niche, marketplace)
            total_saved += saved

    print(f"\n{'='*50}")
    print(f"HARVEST COMPLETE — {total_saved} total products saved")
    print(f"Run: python intelligence/analyzer.py to analyse")
    print(f"{'='*50}")


if __name__ == "__main__":
    main()

File: intelligence/harvester.js
/**
 * Amazon KDP Intelligence Harvester (Node.js) — REFERENCE ONLY
 * ==============================================================
 * WARNING: This script uses raw https.request for scraping.
 * Amazon detects and blocks this approach (CAPTCHA / 503 / empty results)
 * almost immediately on any substantive request volume.
 *
 * THE ACTUAL HARVEST METHOD is the harvester-agent.md agent, which uses
 * Playwright's browser_evaluate() inside a real browser session. This JS file
 * is kept as a reference for the parsing logic only.
 *
 * DO NOT run this script expecting it to successfully harvest Amazon data.
 * Use: `harvest [niche]` to invoke the real harvester-agent via Playwright.
 *
 * Original run intent: node intelligence/harvester.js [niche] [marketplace]
 */

const https = require("https");
const fs = require("fs");
const path = require("path");

const NICHES_PATH = path.join(__dirname, "niches.json");
const OUTPUT_PATH = path.join(__dirname, "harvested.json");

const MARKETPLACES = {
  UK: "www.amazon.co.uk",
  US: "www.amazon.com",
};

const HEADERS = {
  "User-Agent":
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36",
  "Accept-Language": "en-GB,en;q=0.9",
  Accept:
    "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8",
  Connection: "keep-alive",
  "Upgrade-Insecure-Requests": "1",
  "Cache-Control": "max-age=0",
};

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function fetchPage(hostname, path) {
  return new Promise((resolve) => {
    const options = { hostname, path, method: "GET", headers: HEADERS };
    const req = https.request(options, (res) => {
      let data = "";
      res.on("data", (chunk) => (data += chunk));
      res.on("end", () => resolve({ status: res.statusCode, body: data }));
    });
    req.on("error", () => resolve({ status: 0, body: "" }));
    req.setTimeout(15000, () => {
      req.destroy();
      resolve({ status: 0, body: "" });
    });
    req.end();
  });
}

function extractText(html, regex) {
  const m = html.match(regex);
  return m ? m[1].replace(/&#\d+;/g, "").replace(/&amp;/g, "&").replace(/&quot;/g, '"').trim() : null;
}

function parsePrice(text) {
  if (!text) return null;
  const m = text.match(/([\d,]+\.?\d*)/);
  return m ? parseFloat(m[1].replace(",", "")) : null;
}

function parseNum(text) {
  if (!text) return null;
  const m = text.match(/([\d,]+)/);
  return m ? parseInt(m[1].replace(/,/g, ""), 10) : null;
}

function scrapeSearchResults(html) {
  const products = [];
  const seen = new Set();

  // Extract all [data-asin] blocks
  const asinBlocks = html.match(/data-asin="([A-Z0-9]{10})"[^>]*>[\s\S]{0,3000}?(?=data-asin="|<\/div>\s*<\/div>\s*<\/div>\s*<div[^>]*s-result-item)/g) || [];

  // Simpler: find all ASINs and their surrounding context
  const asinPattern = /data-asin="([A-Z0-9]{10})"/g;
  let match;
  while ((match = asinPattern.exec(html)) !== null) {
    const asin = match[1];
    if (seen.has(asin)) continue;
    seen.add(asin);

    // Get context around this ASIN (next 2000 chars)
    const ctx = html.substring(match.index, match.index + 2000);

    // Title
    const titleMatch = ctx.match(/class="a-size-medium[^"]*"[^>]*>\s*([^<]{5,200})</);
    const title = titleMatch ? titleMatch[1].replace(/&amp;/g, "&").replace(/&#(\d+);/g, (_, n) => String.fromCharCode(n)).trim() : null;
    if (!title || title.length < 5) continue;

    // Price
    const priceMatch = ctx.match(/a-offscreen">\s*[£$]([\d,.]+)</);
    const price = priceMatch ? parseFloat(priceMatch[1].replace(",", "")) : null;

    // Rating
    const ratingMatch = ctx.match(/(\d+\.?\d*) out of 5 stars/);
    const rating = ratingMatch ? parseFloat(ratingMatch[1]) : null;

    // Review count
    const reviewMatch = ctx.match(/(\d[\d,]*)\s*(?:rating|review)/i);
    const reviews = reviewMatch ? parseNum(reviewMatch[1]) : null;

    // KU
    const ku = /Kindle Unlimited|a-badge-supplemental/i.test(ctx);

    products.push({ asin, title, price_ebook: price, star_rating: rating, review_count: reviews, on_ku: ku });
  }

  return products;
}

function scrapeProductPage(html, asin) {
  const result = { asin };

  // BSR — look for Best Sellers Rank section
  const bsrSection = html.match(/Best Sellers Rank[\s\S]{0,2000}?(?=<\/ul>|<\/div>|Customers who)/i);
  if (bsrSection) {
    const rankMatches = bsrSection[0].match(/#([\d,]+)\s+in\s+([^<(]+)/g) || [];
    if (rankMatches.length > 0) {
      const first = rankMatches[0].match(/#([\d,]+)\s+in\s+(.+)/);
      if (first) {
        result.bsr_overall = parseNum(first[1]);
        result.bsr_category1 = first[2].trim().replace(/&amp;/g, "&");
      }
      if (rankMatches.length > 1) {
        const second = rankMatches[1].match(/#([\d,]+)\s+in\s+(.+)/);
        if (second) {
          result.bsr_category2_rank = parseNum(second[1]);
          result.bsr_category2 = second[2].trim().replace(/&amp;/g, "&");
        }
      }
    }
  }

  // Review count
  const reviewMatch = html.match(/([\d,]+)\s*(?:global\s+)?ratings?/i);
  if (reviewMatch) result.review_count = parseNum(reviewMatch[1]);

  // Rating
  const ratingMatch = html.match(/([\d.]+) out of 5 stars/);
  if (ratingMatch) result.star_rating = parseFloat(ratingMatch[1]);

  // Kindle price
  const priceMatch = html.match(/kindle-price[\s\S]{0,500}?[£$]([\d,.]+)/i) ||
                     html.match(/Buy now with[^£$]*[£$]([\d,.]+)/i);
  if (priceMatch) result.price_ebook = parseFloat(priceMatch[1].replace(",", ""));

  // KU
  result.on_ku = /kindle.unlimited/i.test(html);

  // Pages
  const pagesMatch = html.match(/([\d,]+)\s*pages?/i);
  if (pagesMatch) result.page_count = parseNum(pagesMatch[1]);

  // Pub date
  const dateMatch = html.match(/(?:Publication date|Date first available)[^:]*:\s*([^<\n]+)/i);
  if (dateMatch) result.pub_date = dateMatch[1].trim();

  return result;
}

async function harvestNiche(niche, marketplace) {
  const host = MARKETPLACES[marketplace];
  const results = [];

  console.log(`\n${"=".repeat(50)}`);
  console.log(`Harvesting: ${niche.name} | ${marketplace}`);
  console.log("=".repeat(50));

  const allProducts = new Map();

  for (const term of niche.search_terms.slice(0, 3)) {
    const encoded = encodeURIComponent(term);
    const searchPath = `/s?k=${encoded}&i=digital-text`;
    console.log(`\n  Searching: "${term}"`);

    const res = await fetchPage(host, searchPath);
    if (res.status !== 200) {
      console.log(`  [!] Got status ${res.status} for "${term}"`);
      await sleep(3000);
      continue;
    }

    const products = scrapeSearchResults(res.body);
    console.log(`  Found ${products.length} products`);
    for (const p of products) {
      if (!allProducts.has(p.asin)) allProducts.set(p.asin, p);
    }

    await sleep(2000);
  }

  console.log(`\n  Total unique: ${allProducts.size}`);
  const productList = Array.from(allProducts.values());

  // Get BSR for top 20
  const top20 = productList.slice(0, 20);
  console.log(`  Getting BSR for ${top20.length} products...`);

  for (let i = 0; i < top20.length; i++) {
    const p = top20[i];
    console.log(`  [${i + 1}/${top20.length}] ${p.asin} - ${(p.title || "").substring(0, 45)}...`);
    const bsrRes = await fetchPage(host, `/dp/${p.asin}`);
    if (bsrRes.status === 200) {
      const bsrData = scrapeProductPage(bsrRes.body, p.asin);
      Object.assign(p, bsrData);
      console.log(`    BSR: ${p.bsr_overall || "N/A"} | Reviews: ${p.review_count || "N/A"} | KU: ${p.on_ku || false}`);
    } else {
      console.log(`    [!] Could not fetch product page (status: ${bsrRes.status})`);
    }
    await sleep(1500);
  }

  for (const p of productList) {
    results.push({
      ...p,
      niche: niche.name,
      marketplace,
      scraped_at: new Date().toISOString(),
    });
  }

  return results;
}

async function main() {
  const niches = JSON.parse(fs.readFileSync(NICHES_PATH, "utf8")).niches;
  const targetNiche = process.argv[2] || null;
  const targetMarketplace = process.argv[3] || null;

  let allResults = [];

  // Load existing results if any
  if (fs.existsSync(OUTPUT_PATH)) {
    allResults = JSON.parse(fs.readFileSync(OUTPUT_PATH, "utf8"));
    console.log(`Loaded ${allResults.length} existing records`);
  }

  for (const niche of niches) {
    if (targetNiche && niche.name !== targetNiche) continue;
    for (const marketplace of niche.marketplaces) {
      if (targetMarketplace && marketplace !== targetMarketplace) continue;
      const results = await harvestNiche(niche, marketplace);
      allResults = allResults.filter(
        (r) => !(r.niche === niche.name && r.marketplace === marketplace)
      );
      allResults.push(...results);
    }
  }

  fs.writeFileSync(OUTPUT_PATH, JSON.stringify(allResults, null, 2));
  console.log(`\n${"=".repeat(50)}`);
  console.log(`HARVEST COMPLETE — ${allResults.length} total products`);
  console.log(`Saved to: ${OUTPUT_PATH}`);
  console.log("=".repeat(50));
}

main().catch(console.error);
